dvc/repo/gc.py

```diff
@@ -134,11 +135,25 @@ def gc(  # noqa: PLR0912, PLR0913, C901
     if not cloud:
         return
 
-    for remote_odb, obj_ids in odb_to_obj_ids:
-        if remote_odb is None:
-            remote_odb = repo.cloud.get_remote_odb(remote, "gc -c")
+    for remote_odb, obj_ids in odb_to_obj_ids.items():
+        assert remote_odb is not None
         removed = ogc(remote_odb, obj_ids, jobs=jobs)
         if removed:
             get_index(remote_odb).clear()
         else:
             logger.info("No unused cache to remove from remote.")
+
+
+def _merge_remote_obj_ids(
+    repo: "Repo", remote: Optional[str], used_objs: "ObjectContainer"
+):
+    # Merge default remote used objects with remote-per-output used objects
```

Comment:
This is not 2.x/legacy cache specific, but I'm think `gc -c` was probably broken before this when using per-output-remotes in the event that either the default remote or the `-r <remote>` argument matched a per-output-remote.

In that scenario we would previously end up trying to gc the remote twice, once with the set of used-objects for the default/`-r` case and once with the set of used-objects for the per-remote case. This meant that in the default/`-r` iteration we would gc any objects that were considered used with the `remote:` field, and in the per-remote iteration we would gc any objects that did not specify the `remote:` field.

This does not apply to push/fetch which is probably why we never identified the issue before, but for gc in particular we have to make sure the we pass all possible used object IDs for a given ODB when we call `odb.gc()`.

In the push/fetch case it doesn't make a difference (other than efficiencey) if we do `odb.push(some_objs)` and then another separate `odb.push(more_objs)`. But in the gc case this causes a loss of data unless we only make a single `odb.gc(some_objs + more_objs + ...)` call.